### DISCUSSION TITLE
subscriptions-transport-ws seems to have global state, multiple clients with subscription clients use last created client

### DISCUSSION THREAD

**Discussion by bartenra** [urql-graphql/urql#1814] [Q&A]
Just a PSA that `subscriptions-transport-ws` seems to have some global state.

I instatiate two clients each with a `subscriptionExchange` with the wrapper function that can be found below:

```ts
const client1 = urqlClient("token1");
const client2 = urqlClient("token2");
```

When I then use `client1.subscribe(...)`, `token2` is used. If instead I define the clients the other way around.

```ts
const client2 = urqlClient("token2");
const client1 = urqlClient("token1");
```

And call `client1.subscribe(...)` then `token1` is correctly printed.

My use-case for defining two clients here is that I am writing tests for subscriptions...

The way I get around this is by creating a new subscription client inside `forwardSubscription`. 

---

```ts
import { SubscriptionClient } from 'subscriptions-transport-ws';
import { Client, dedupExchange, fetchExchange, subscriptionExchange } from '@urql/core';
import { cacheExchange } from '@urql/exchange-graphcache';

export const urqlClient = (token: string) => {
  const subscriptionClient = new SubscriptionClient(`ws://localhost:8000/graphql/?token=${token}`, { reconnect: false }, WebSocket);
  let exchanges = [
    dedupExchange,
    cacheExchange(cacheExchangeSettings),
    subscriptionExchange({
      forwardSubscription(operation) {
        console.log("forwarding subscription ", token);
        return subscriptionClient.request(operation);
      },
    }),
    fetchExchange
  ];

  return new Client({
    url: 'http://localhost:8000/graphql/',
    fetchOptions: {
      'headers': {
        "Authorization": `JWT ${jwtToken}`
      }
    },
    exchanges
  })
}
```

**Comment by JoviDeCroock** [comment]
Hey, that's a good find I guess you could also try using [`graphql-ws`](https://github.com/enisdenjo/graphql-ws) if it's still in your realm of possibilities. Quickly scanned the `subscription-transport-ws` code and couldn't seem to find any global variables so might be a bug worth mentioning there.

**Reply by bartenra** [reply]
Yes that would be ideal since `subscriptions-transport-ws` is also deprecated.

However, we use [Ariadne](https://ariadnegraphql.org/) as a backend, which uses the same protocol used by `subscriptions-transport-ws`, `graphql-ws`' protocol is slightly different if I am not mistaken.
